spec/models/user_spec.rb

```diff
@@ -40,94 +44,216 @@
+    end
+    it 'returns false when user belongs to no groups' do
       allow(subject).to receive(:groups).and_return([])
-      expect(subject.is_viewer).to be_falsey
-      expect(subject.is_viewer?).to be_falsey
+      expect(subject.solr_role_allowed?(solr_doc, 'roleA')).to be false
     end
   end
 
   describe 'roles' do
-    before(:each) do
-      @answer = {}
-      @doc = {
-        'apo_role_dor-administrator_ssim' => ['workgroup:dlss:groupA', 'workgroup:dlss:groupB'],
-        'apo_role_sdr-administrator_ssim' => ['workgroup:dlss:groupA', 'workgroup:dlss:groupB'],
-        'apo_role_dor-apo-manager_ssim'   => ['workgroup:dlss:groupC', 'workgroup:dlss:groupD'],
-        'apo_role_dor-viewer_ssim'        => ['workgroup:dlss:groupE', 'workgroup:dlss:groupF'],
-        'apo_role_sdr-viewer_ssim'        => ['workgroup:dlss:groupE', 'workgroup:dlss:groupF'],
-        'apo_role_person_dor-viewer_ssim' => ['sunetid:tcramer'],
-        'apo_role_person_sdr-viewer_ssim' => ['sunetid:tcramer'],
-        'apo_role_group_manager_ssim'     => ['workgroup:dlss:groupR']
+    # The exact DRUID is not important in these specs, because
+    # the Dor::SearchService is mocked to return solr_doc.
+    let(:druid) { 'druid:ab123cd4567' }
+    let(:answer) do
+      {
+        'response' => { 'docs' => [solr_doc] }
       }
-      @answer['response'] = { 'docs' => [@doc] }
-      allow(Dor::SearchService).to receive(:query).and_return(@answer)
-      @user = User.find_or_create_by_webauth(double('webauth', :login => 'asdf'))
     end
-    it 'should build a set of roles' do
-      expect(@user).to receive(:groups).and_return(['workgroup:dlss:groupF', 'workgroup:dlss:groupA'])
-      expect(@user.roles('pid')).to eq(['dor-administrator', 'sdr-administrator', 'dor-viewer', 'sdr-viewer'])
+    let(:solr_doc) do
+      {
+        'apo_role_sdr-administrator_ssim' => %w(workgroup:dlss:groupA workgroup:dlss:groupB),
+        'apo_role_sdr-viewer_ssim'        => %w(workgroup:dlss:groupE workgroup:dlss:groupF),
+        'apo_role_dor-apo-manager_ssim'   => %w(workgroup:dlss:groupC workgroup:dlss:groupD),
+        'apo_role_person_sdr-viewer_ssim' => %w(sunetid:tcramer),
+        'apo_role_group_manager_ssim'     => %w(workgroup:dlss:groupR)
+      }
+    end
+    before(:each) do
+      allow(Dor::SearchService).to receive(:query).and_return(answer)
+    end
+    it 'should accept any object identifier' do
+      expect{subject.roles(druid)}.not_to raise_error
+      expect{subject.roles('anyStringOK')}.not_to raise_error
+    end
+    it 'should return an empty array for any blank object identifer' do
+      ['', nil].each do |pid|
+        expect{subject.roles(pid)}.not_to raise_error
+        expect(subject.roles(pid)).to be_empty
+      end
+    end
+    it 'should build a set of roles from groups' do
+      doc_groups = %w(workgroup:dlss:groupF workgroup:dlss:groupA)
+      doc_roles = %w(sdr-administrator sdr-viewer).sort
```

Comment:
:ok_hand: -- it had a `doc` prefix because I was thinking of it related to Solr content, but I agree with the change.